### ISSUE TITLE
Commands enhancements and threading

### ISSUE THREAD

**Issue by Siegen** [Bionus/imgbrd-grabber#621]
I have a few ideas for making the Commands more usable for more database structures:
1. Allow the tag and image portions of the Commands and Database settings to be reversible. Either a setting to switch the order or something like tags_before fields and tags_after fields.
2. Add a metadata token to Image commands for typed tags (tag categories). %allo% and %all% are untyped.
3. Add the image metadata fields to the Tag commands. If nothing else, %md5% or %path% would be hugely helpful in allowing the tag commands to find the image in the database without resorting to fragile tricks like MAX(id) that can break with the bulk download.

I bring these items up because it seems to be an intractable problem when trying to upload type-ed tags to shimmie. Shimmie has its tags and images in different tables (unlike gelbooru) and expects each id to exist before it will allow rows in the image_tags table (precluding tricks like MAX(id) + 1).

If you upload the tags through a tag command (the only way to get them with their type), you get stuck trying to associate them with the image, which comes after. If you try to let the image (or the After command) associate the tags with the image, you don't have the %type%s anymore. Which means you can't guarantee you won't mistake identical tags with different categories (ie. Naruto the copyright vs Naruto the character). 

And anything done in the Image or Before commands doesn't have access to the %type%s so you can only upload untyped tags (or, as above, risk ambiguities in trying to divine the type from existing tags).

The best solution I've found is to have a staging area image already in shimmie (just a junk image without any tags), use the Tag command to set all the typed tags to that known image's id, then let the Image or After command update all the tags from that junk image's id to the MAX(id) from the images table after the new image has been added.

That works okay when you're going slowly enough (and we assume there's only ever one user accessing the database) but breaks during bulk downloads. It's also pretty messy, anyways. It seems like letting all of the command fields use any of the metadata tokens shouldn't be impossible and that there should be an option for dealing with the tags after dealing with the image (that seems more logical, in my mind).

**Comment by Bionus** [comment]
Hello

> Allow the tag and image portions of the Commands and Database settings to be reversible. Either a setting to switch the order or something like tags_before fields and tags_after fields.

You mean to be able to choose between saving the tags before or after the image?

Or have like three commands fields:
- Tags (before)
- Image
- Tags (after)

> Add a metadata token to Image commands for typed tags (tag categories). %allo% and %all% are untyped.

This has been included in the 4.7.0 update, in the shape of the [Filename options](https://github.com/Bionus/imgbrd-grabber/wiki/Filename#options) (and more especially the `includenamespace` option). What you need for this kind of bulk upload is `%all:includenamespace,unsafe%` (I advice against using `%allo%` unless necessary, as I'm likely to deprecate it).

> Add the image metadata fields to the Tag commands. If nothing else, %md5% or %path% would be hugely helpful in allowing the tag commands to find the image in the database without resorting to fragile tricks like MAX(id) that can break with the bulk download.

Why not, there's no problem with letting tag commands access all fields.

BTW: could you share the SQL commands you use for Shimmie? Maybe I can add another tutorial in the wiki to configure it, although I don't see why would someone choose Shimmie over MyImouto (?).

**Comment by Siegen** [comment]
Yeah, either an option to switch the order of operations so tags are processed after images (how shimmie would prefer it) or an extra set of commands for tags that process after the image (the same loop, just after the image processing).

> This has been included in the 4.7.0
Ah, I completely missed the namespace option, somehow. I'll take another swing at letting the Image command handle tags then.

> I don't see why would someone choose Shimmie over MyImouto
https://github.com/myimouto/myimouto/issues/204
The owner said he'd be deleting the repository (for some reason). Shimmie has been around a while and the maintainers don't seem to want to obliterate the codebase.

I'll check out MyImouto but, unless the feature set is vastly better than Shimmie, I'd prefer to use a project that isn't a fork of a delete repository of a port from another language.

> could you share the SQL commands you use for Shimmie
Sure thing. Once I get the namespace working, it should be solid. Though, I'm currently using a combination of the SQL commands and a php script to make life easier.

**Comment by Siegen** [comment]
Okay, so the namespace solved my issues. Works great with bulk download, now.

As for what I'm doing:

1. Grabber needs to have "Replace spaces by underscores" on and a non-space tag delimiter. This setup also assumes the user has php installed and set up in their PATH. If Shimmie is installed on a remote server, the php file could be placed there and called via curl or the like but the argument vector would have to be switched out for $_REQUEST variables.

2. Commands > Image = `php "UploadToShimmie.php" -h="%md5%" -f="%path%" -t="%all:includenamespace,unsafe%" -s="%website%" -r="%rating%"`

3. Place attached file in Grabber install directory and rename to UploadToShimmie.php
[UploadToShimmie.example.txt](https://github.com/Bionus/imgbrd-grabber/files/490317/UploadToShimmie.example.txt)

4. In UploadToShimmie.php:
- Set $uploadURL to your Shimmie's upload form.
- Set the database settings for your Shimmie database and user.
- Set $tagDelimiter to whatever is set in Grabber under Save > Tags separator.
- Other settings can be tweaked as desired.

--

Ultimately, I decided to move everything into the php file. It provides a lot of flexibility that I wanted for things like category replacement. The basics could be moved to the SQL Commands in Grabber but you'd lose at least the thumbnail generation on top of interacting with the extensions and doing other custom processing.

If you do want to include the SQL Commands, though:
Image: 
`INSERT INTO
    images
    (
        owner_id,
        owner_ip,
        filename,
        filesize,
        hash,
        ext,
        source,
        width,
        height,
        posted
    )
VALUES
    (
        2,
        "::1",
        "%md5%.%ext%",
        1,
        "%md5%",
        "%ext%",
        "%source%",
        %width%,
        %height%,
        NOW()
    )`
(Replace "%md5%.%ext%" with whatever your naming convention is)

Tag:
`INSERT INTO
    tags
    (
        tag, 
        count
    )
VALUES
    (
        REPLACE('%type%:%tag%', 'general:', ''), 
        1
    )
ON DUPLICATE KEY 
    UPDATE 
        count = count + 1;

INSERT INTO 
    image_tags
    (
        image_id,
        tag_id
    )
SELECT
    16,
    tags.id
FROM
    tags
WHERE
    tags.tag = REPLACE('%type%:%tag%', 'general:', '');`
(Can remove the REPLACE call, if you don't want to strip out the general: tag type)

After:
`UPDATE
    image_tags
SET
    image_id = (SELECT MAX(images.id) FROM images)
WHERE
    image_id = 16;
    
DELETE FROM
    image_tags
WHERE
    image_id =16;`

You'd also need to add this to the Image Command:
`CopyImage.bat "%path%" "%md5%"`

And place a bat file named CopyImage.bat in the Grabber install dir with these contents:
`SET _dest=%2
SET _destDir=%_dest:~0,2%

mkdir P:\wamp\www\shimmie\images\%_destDir%

"C:\Windows\System32\cmd.exe" /C move %1 "P:\wamp\www\shimmie\images\%_destDir%\%_dest%"`
(Replace P:\wamp\www\ with whatever your shimmie install path is)

That bat file will copy the saved image from wherever Grabber put it to the correct place it should be in Shimmie (...\images\ab\%md5% where ab are the first two characters of the md5 hash).

Like I said, though, that will only work if the user is saving images with a meaningful delay between them. It'd probably also mess up if the user tried to save an image that was already in shimmie, since the tags would then probably get set on the previously saved image.

**Comment by Siegen** [comment]
Ah, I may have spoke too soon. The mass downloader now seems to be setting every tag type as "unknown" while the manual downloader is working fine. This appears to be happening regardless of the source website or the search query. 

I can see in the commands.log that it's grabber that's passing the tag types as "unknown" and not a bug later on in the pipeline.

**Comment by Siegen** [comment]
Yeah, the example file just has that one category being stripped but my production file has several categories being stripped or replaced. In all, I'm mutating near-about a thousand tags so I wanted a lot of explicit control over how the categories were routed, including bulk category replacement based on tag presence.

The separator parameter is a good trick, though.

Here's an updated PHP file with more features and a few bug fixes.
[UploadToShimmie.example.txt](https://github.com/Bionus/imgbrd-grabber/files/504802/UploadToShimmie.example.txt)

Thank you for adding those new features.

**Comment by Bionus** [comment]
>  If you'd like any help with the wiki, let me know. Shimmie is dead-simple to install but there are a few settings that people should probably set if they want a workable experience.

I just finished creating the page: https://github.com/Bionus/imgbrd-grabber/wiki/Shimmie
I simplified a few steps (using the separator and the full path to PHP) but in general I just used your guide you commented here, thanks very much for it. It's still quite verbose because I wanted to make it as clear as possible for non-tech people.

If you have any suggestions, find any mistakes, or just want to change the `UploadToShimmie.php` file, feel free to either edit it or comment here! I also added you in the header as thanks for your help, but if you don't want your name to appear you can also remove it. :smile:

Note that I made a small edit in the PHP file to work with PHP 5.6 line 186. I also added a part to allow uploads bigger than 1MB.

**Comment by Siegen** [comment]
> Note that I made a small edit in the PHP file to work with PHP 5.6 line 186

Oh, right. I forgot about PHP's dislike of anonymous objects. Been dealing with C# for too long.

> I also added a part to allow uploads bigger than 1MB.

That was going to be my suggestion for making shimmie workable. Though, people might also want to increase the thumbnailer memory limit, if they download very large images.

My only comment on the page is that people can edit the shimmie settings under Admin > Board Config and the PHP settings by left-clicking the WAMP tray icon, then clicking on PHP > PHP Settings > max_post_size. Could be easier than doing it manually.
They can also increase the upload_max_filesize and memory_limit settings while they're there, if they want.

> or just want to change the UploadToShimmie.php file

I'll probably get a version working with remote servers done today or tomorrow. Should be identical except for removing the need to change Shimmie's anonymous user class.

**Comment by Bionus** [comment]
> My only comment on the page is that people can edit the shimmie settings under Admin > Board Config

The problem is that the current tutorial does not show how to have an admin account (and I have no idea either). Is there one by default, like "admin"/"admin"?

> and the PHP settings by left-clicking the WAMP tray icon, then clicking on PHP > PHP Settings > max_post_size. Could be easier than doing it manually.

Yeah, I agree, but since my WAMP is half-broken these shortcuts doesn't work for me, so I couldn't guarantee that they work, unfortunately. 😢

> I'll probably get a version working with remote servers done today or tomorrow. Should be identical except for removing the need to change Shimmie's anonymous user class.

Nice! That would simplify a few steps and allow more complex configurations then. Feel free to tell me if you want me to update the one in the wiki. Else, it's simply located in the "attachments" directory of the [wiki's git](https://github.com/Bionus/imgbrd-grabber.wiki.git).

**Comment by Siegen** [comment]
> The problem is that the current tutorial does not show how to have an admin account (and I have no idea either). Is there one by default, like "admin"/"admin"?

After the install, it tells you that the first account you make is made admin automatically. So, the user would land at the index after install, click Create Account, and then use that as their admin account.

**Comment by Siegen** [comment]
Okay, much trial and tribulation was had.

Here's the new php file. This works for both local and remote servers. The server settings just need to be set properly. The default settings here are for a local server.
[UploadToShimmie.example.txt](https://github.com/Bionus/imgbrd-grabber/files/505357/UploadToShimmie.example.txt)

It now properly logs into Shimmie with the specified user account so the instructions about changing the user class file can be removed. Consequently, the user will *have* to make a user account before using this script so they'll have the admin account you were worried about above.

I don't know if it's outside the scope of the wiki but people will need to make sure their remote host allows mysql traffic (tcp on port 3306 or w/e the mysql port is set to), if they want to manually issue any queries (such as for the ratings extension). For ubuntu on my host, this meant setting the "bind-address" field in /etc/mysql/mysql.conf.d/mysqld.cnf to an internet-facing interface, rather than the default loopback address. That was somewhat less-than-obvious but might be specific to my setup so I'll let you decide how much detail you want to give.
As well as the usual firewall stuff.

It might be helpful to give a troubleshooting tip of opening the commands.log file in the grabber settings dir, copying the last command (that presumably failed), opening a command prompt, cd'ing to the grabber install directory, and pasting that command in to see what error php throws out.

**Comment by Siegen** [comment]
Here's a version that won't needlessly die if the person doesn't have mysql accessible: 
[UploadToShimmie.example.txt](https://github.com/Bionus/imgbrd-grabber/files/507406/UploadToShimmie.example.txt)

I was thinking, you might want to add an option to execute the Commands asynchronously. Not sure if it works in linux to call the command asynchronously but "start" doesn't seem to work for windows, when called from the Image Command. I had to make a bat file that I call from the Command which itself calls "start php". Not sure how a large batch download would like it but saving images manually becomes a bit tedious with a network upload since the entire UI waits for the Image Command to return.

Splitting Grabber's GUI thread from its logic thread would probably be a more robust solution but idk how much work that'd be for your codebase.

Just an idea.

**Comment by Bionus** [comment]
> Here's the new php file. This works for both local and remote servers. The server settings just need to be set properly. The default settings here are for a local server.

Nice, even though I'm pretty sure most people will use Grabber with a local server.

> I don't know if it's outside the scope of the wiki but people will need to make sure their remote host allows mysql traffic (tcp on port 3306 or w/e the mysql port is set to), if they want to manually issue any queries (such as for the ratings extension). For ubuntu on my host, this meant setting the "bind-address" field in /etc/mysql/mysql.conf.d/mysqld.cnf to an internet-facing interface, rather than the default loopback address. That was somewhat less-than-obvious but might be specific to my setup so I'll let you decide how much detail you want to give.
> As well as the usual firewall stuff.

Not really in scope. Local setup, being the main use case is described in details, but for power users wanting to call a remote server, it's up to them. I'll add a small *post scriptum* or example for mysql though, since you already have the procedure.

> It might be helpful to give a troubleshooting tip of opening the commands.log file in the grabber settings dir, copying the last command (that presumably failed), opening a command prompt, cd'ing to the grabber install directory, and pasting that command in to see what error php throws out.

I could add it in the Commands wiki page, but users either won't be able to do it (casual users) or already know it (power users).

About threading, that would be easy to implement, especially for "normal" commands (not SQL), since all there is to do is pop a thread off some thread pool and run the command in it in the background. As for SQL, that would either require some lock on the connection or to have one SQL connection for every thread, which are both do-able.
The problem is that since the order of the commands (for a single image) has to be guaranteed to be "start >tags (before) > image > tags (after) > end" you can't pop one thread for every command but one thread by image. And even if you don't block the UI by using a thread, it won't allow parallelism (except for the concurrent downloads thing) since it could make the threads late if your commands are slower than your downloads: the downloader will wait for the thread to finish before starting the next image.

**Comment by Siegen** [comment]
Having one worker thread would probably be enough. The point of my recommendation is quality-of-life, UX stuff, rather than throughput improvements.

That is, I have it now running the Image Command using "start" which doesn't make the uploads too much faster (there's only so much bandwidth between me and my server anyways) but does let the UI continue on its merry way. So, instead of opening 10 images and slowly clicking "Save & Close" on each one with a multi-second delay between each, I can click through each one as quickly as the UI will allow and the uploads progress in the background.

At that point, you could just shove all the save requests in a queue on the worker thread and use the existing code to go through each in order.

Not that I'd argue against doing further parallelization with a thread pool but it sounds like it'd require more refactoring than a queue on one thread.

Just spitballing, anyways.

**Comment by Bionus** [comment]
Having a thread pool with as least as many threads as concurrent image downloads (max 10) would allow parallelization of commands at the same cost as running a single one (I believe it would even be more easy to use a thread pool as using a single thread would block other commands until his is executed, requiring to queue them, which would result in some kind of 1-thread thread pool). It would also allow to move the actual image download and saving to that thread easily (since it's got nothing to do in the UI thread either).

> As well, the real point is that it's generally considered good design to have the UI and logic running on different threads so these kinds of unresponsive UI problems don't happen. But that's probably a bigger project than just having a thread dedicated to saving images.

Indeed, however the current state of the downloader was designed 5 years ago, and as that time it was more of a POC than anything (plus my skills where far from now), so this kind of things I just ignored. That's why currently, the whole download part of the program is closely coupled with the UI, which makes it impossible in its current state to move to another thread. The only threaded stuff is image loading from byte array. 😞

That's actually the main refactors I'm working on at the moment, where I'm moving more and more of the program's logic out of the UI objects (windows and dialogs), first in order to add unit tests, and then to refactor/threading the logic itself when I'm sure no breaking changes can occur. That's the main reason I created the Downloader class a while back (and many other classes), which could be easily moved to another thread, but currently only manages the pages' download at the beginning of batches.

> The point of my recommendation is quality-of-life, UX stuff, rather than throughput improvements.

Is there other places except for "image window > save" and batch downloads where slowdowns happen? (while I'm at it, so I can fix them all at the same time)
